### PR TITLE
sql: add support for date_trunc(string, interval)

### LINKED ISSUE #76960: sql: add support for date_trunc(string, interval)
After a [recent commit](https://github.com/django/django/commit/b7f263551c64e3f80544892e314ed5b0b22cc7c8), Django 4.1 (to be released in August 2022) will generate SQL like:
```sql

SELECT EXTRACT('second' FROM DATE_TRUNC('second', "duration"))
```
where `duration` is an `interval` column. This works on PostgreSQL but not on CockroachDB as of the latest nightly v22.1.0-alpha.1-1607-g1f4dde2c5f @ 2022/02/24:
`psycopg2.errors.UndefinedFunction: extract(): unknown signature: date_trunc(string, interval)`.

#43209 was a similar issue adding support for `extract(string, interval)`.

### PR DESCRIPTION
**Pull request by ecwall**
fixes https://github.com/cockroachdb/cockroach/issues/76960

Release justification: Adds new date_trunc interval support that does not affect
existing functionality.

Release note (sql change): Add support for date_trunc(string, interval).

### CODE REVIEW

File: pkg/sql/sem/builtins/builtins.go [hunk 1]
```diff
@@ -8715,6 +8729,69 @@ func truncateTimestamp(fromTime time.Time, timeSpan string) (*tree.DTimestampTZ,
 	return tree.MakeDTimestampTZ(toTime, time.Microsecond)
 }
 
+func truncateInterval(fromInterval *tree.DInterval, timeSpan string) (*tree.DInterval, error) {
+
+	toInterval := tree.DInterval{}
+
+	switch timeSpan {
```
Comment (otan) [inline pkg/sql/sem/builtins/builtins.go:8736]: nit: can we move strings.ToLower to here?

File: pkg/sql/logictest/testdata/logic_test/interval [hunk 2]
```diff
@@ -650,3 +650,96 @@ WHERE tablename = 'intervalstyle_in_index'
...
+# date_trunc
+statement ok
+SET intervalstyle = 'postgres'
+
+statement error pgcode 22023 interval units "invalid" not recognized
+SELECT
+  date_trunc('invalid', interval '1 month')
+
+query TTT
+SELECT
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/interval:663]: optional, but if you wanted to make this table driven-y for non-errors, you could make it:

```
query TTT
SELECT i, truncate_str, date_trunc(truncate_str, i) FROM (VALUES
   ('decade', interval '1001 year 1 month 1 day 1 microsecond'),
   ('millennium', '1001 year 1 month 1 day 1 microsecond')
   -- etc
) AS t(truncate_str, i)
```

**Review by otan — APPROVED** [review]

File: pkg/sql/sem/builtins/builtins.go (hunk 1 above)
Comment (ecwall) [reply pkg/sql/sem/builtins/builtins.go:8736]: The other similar truncate/extract function call sites all are doing `strings.ToLower` before calling the function.  If we do it for this one I think we should do it for all of them, but it should probably be in a separate PR.

**Review by rafiss — COMMENTED** [review]
in case it's helpful for checking on any other odd edge cases, here's the postgres source: https://github.com/postgres/postgres/blob/45fb0de4dc65f43a037fe7c90f360ae0596d9328/src/backend/utils/adt/timestamp.c#L4301

**Comment by ecwall** [comment]
> in case it's helpful for checking on any other odd edge cases, here's the postgres source: https://github.com/postgres/postgres/blob/45fb0de4dc65f43a037fe7c90f360ae0596d9328/src/backend/utils/adt/timestamp.c#L4301

Yeah it looks like this confirms that only `'week'` has special handling.

**Comment by ecwall** [comment]
bors r=otan

**Pull request merged**